niri-support

clotodex opened the pull request:
Looking for testers! If you are on nixos, use this url for ashell: `ashell.url = "github:clotodex/ashell-iwd/feature/niri-support";`
- created a new CompositorService
- implemented niri support for:
  - [x] workspaces
  - [x] window title
  - [x] keyboard layout (module still todo)
- pulled hyprland implementation out of the modules into the service
- The CompositorService will first try and load hyprland - if that fails it will load niri


I ran some inital tests on my machine to validate it working, but I did not put it through the ringer yet:
- [x] verify colors of the workspaces
- [x] test (and fix) commands -> scrolling and clicking does not work yet (broken for both)
  - [x] this also requires forwarding the commands to the chosen backend 
- [x] test hyprland

**Open decisions:**
- [x] should the backend be something to be "failed over to" like it is right now and with network, or a config option?
- [x] Readme, intro, website, doc adjustments - this is no longer just limited to hyprland :)
- [x] debate empty / full init state of the service - right now niri immediately fills it, so would like to keep it empty
- [x] changed the scrolling on worksapces to be in the right direction - was this a bug or intended? I assumee it was "if I scroll up I expect the workspace number to increase" while scrolling layouts (also on hyprland) scroll down with increasing workspace index. This will depend on how you configure workspace animation direction and might just be a config preference - or something to be handled on a compositor to compositor basis - what are your takes?

Problems that remain open (but would merge anyway, potentially with doc comment):
- niri can lead to an infinite loop with enable_workspace_filling, since niri removes empty ones - should be checked or mentioned somehow
- there is no submap for niri atm https://github.com/YaLTeR/niri/issues/846 and https://github.com/MaxVerevkin/wlr-which-key (which is what most are using I assume) does not have a dbus / ipc

This also allows implementing other compositors and wayland porotocols later.
*(I did not implement the ext/workspaces yet - I could not get the smithay client to work on yielding the right events (the control loop is different - tried to copy idle_inhibitor but failed so far)
I focused on a full ipc implementation, since wayland afaik has no way of discovering window titles atm - so an implementation was needed anyways*

Any feedback, especially on architectural choices (@MalpenZibo ) are very welcome!

clotodex commented on src/services/compositor/niri.rs:
```diff
@@ -0,0 +1,252 @@
+    stream.write_all(json.as_bytes()).await?;
+    stream.flush().await?;
+
+    let mut reader = BufReader::new(stream);
```
not ready yet, but cool reference code

jazzpi commented:
> I would love to have it as a fall back compositor implementation for sure. In terms of what it exposes and how we deal with it - i am already fighting with niri, since it does not expose any kind of monitor numbering - nor has an idea of what a primary_monitor is. So I will have to figure out some hacks or defaults.

With the workspace protocol we can map workspaces (`ext_workspace_handle_v1`) to outputs (`wl_output`) in a compositor-agnostic way. `wl_output`s also don't have a monitor numbering though, but they do have names and descriptions. As far as I can see, the monitor ID is only used for the workspace colors, right? That could be mapped from the name or description instead (but that would be a breaking change of course).

> This PR will likely still move a lot in its API. Right now my plan would be to clean this up and merge it - then add new compositor implementations as PRs on top. If this is too slow for you, let me know and we can somehow merge earlier.

I think this PR is much further along than my branch, so I'd rather base the workspace protocol implementation off this :)

MalpenZibo commented:
I will look as soon as possible. Thank you so much, this is very appreciated. Sadly, I haven't had much free time lately, but I'll try my best to check this out!

clotodex commented:
> I will look as soon as possible. Thank you so much, this is very appreciated. Sadly, I haven't had much free time lately, but I'll try my best to check this out!

To help the review:

What I did:
- pulled out hyprland into the CompositorService
- implemented niri using the niri-ipc
- rewrote the affected modules to use the compositorservice

How to find yourself around the files:
- start with types.rs to see the new interface / message the service can expose
- then move on to compositor/mod.rs
- the CompositorService has a single Message which is CompositorStateChanged
- there is also an ActionPerformed Message that gets triggered e.g. on workspace click. This is useful if in the future a compositor needs a hint to recompute - with the IPC this is a noop.
- the service keeps track of its `CompositorChoice`, currently only hyprland and niri and otherwise delegates its work to the specific implementation
- the modules now only listen to the same service subscription and consume the state - this also needed changes in message communication since the commands cant be performed only within a module anymore (hyprland specific code was there before)

Hope this helps :)

What to look out for especially:
- I see two risks: 1) in moving out hyprland I could have made some mistakes and lost edge cases (I tested everything but still)
- and 2) In pulling the hyprland-specific logic from the modules into the services the same applies (again also tested)
- I feel like I made a few tradeoffs to get this working soon, leaving some architectural choices for later => anything you think we should resolve now, let me know

Otherwise looking forward to see what I don't :)

MalpenZibo commented:
About 

> **Open decisions:**
> 
>     * [ ]  should the backend be something to be "failed over to" like it is right now and with network, or a config option?
> 
>     * [ ]  Readme, intro, website, doc adjustments - this is no longer just limited to hyprland :)
> 
>     * [ ]  debate empty / full init state of the service - right now niri immediately fills it, so would like to keep it empty
> 
>     * [ ]  changed the scrolling on worksapces to be in the right direction - was this a bug or intended? I assumee it was "if I scroll up I expect the workspace number to increase" while scrolling layouts (also on hyprland) scroll down with increasing workspace index. This will depend on how you configure workspace animation direction and might just be a config preference - or something to be handled on a compositor to compositor basis - what are your takes?

- It would be great to have some sort of env var to know which compositor is in use, but probably it's not so easy, so we could leave it as it is for now.
- Readme and other stuff can be done in a dedicated PR.
- I still need to see the code. Sorry, I haven't had enough free time recently :cry:   
- I don't know really XD I personally never use the scrolling feature. Maybe @EdgesFTW could help us.

clotodex commented:
> > Since they are all subscribing to the same compositor service, this is actually just a single subscription (ice deduplicates them)
> 
> Yes but I made a simple test, ...
> 
> And I notice that with two open windows, moving the mouse between the two causes two `"event hyprland"` log using both `workspaces` and `window_title` module, and only one using only the `window_title` module. So, there's something I don't understand. Could you try to see if it's something that happens also on your side?

It is actually not true that iced deduplicates the subscription - it might be a race condition though. In my logs:
```
WARN [ashell::services::compositor] Failed to listen to hyprland: Could not get socket path! (Is Hyprland running??)
WARN [ashell::services::compositor] Failed to listen to hyprland: Could not get socket path! (Is Hyprland running??)
WARN [ashell::services::compositor] Listening for niri instead
WARN [ashell::services::compositor] Listening for niri instead
```
Each should obviously only happen once. Will read up on iced a bit.

clotodex commented:
Alright so summarizing
- i will implement the env var failover to auto-detect what compositor is in use - instead of relying on errors
- we do the renaming/docs as a separate cleanup pr
- the init state of the service stays as is - no errors from few weeks of use
- scrolling stays as implemented until we hear back from @EdgesFTW 
- i will pull the subscription to app level (will check the fancyness when coding)
- any extension - wayland protocols, submaps, etc stay for future work

Other than that, any major annoyances you have with the code so far?

clotodex commented:
Alright. The multi-consumer situation is a bit cursed. Currently the setup is like this:

module1 -> subscribes service -> service subscribes to broadcast -> niri/hypr start listening and manage state -> state passed back -> message triggered in module 1 -> module1 updates service instance -> service instance state changes
The problem: with a module2, only the internal listening is shared - state is managed a second time, init etc needs to be carefully managed.

This leads me to two paths:
1. As @MalpenZibo suggested - pull everything to app, to specialized message passing - always subscribe to compositor IPC (which you likely always want) - manage all submodules via specialized messages managed in the service **main pro: simplicity and effeciency, main con: VERY tight coupling and putting the work to modules**
2. Keep the broadcasting idea, split the service. We could have a Workspaces Service, a Window Title Service, etc - all listening to the same IPC broadcast and constructing their states independently. **main pro: services manager their own responsibility, fully-declarative by modules, subscription and UI code separated nicely per module, main con: state duplication, potentially no real gain**

MalpenZibo commented on src/modules/workspaces.rs:
```diff
@@ -163,161 +168,162 @@ fn get_workspaces(config: &WorkspacesModuleConfig) -> Vec<Workspace> {
+                        .command(CompositorCommand::ScrollWorkspace(direction))
+                        .map(Message::ServiceEvent);
+                }
+                return iced::Task::none();*/
```
ok, we could use the ScrollWorkspace command but we should add a way to check if the scroll arrives from the touchpad and filter the sensitivity.

It could be done in a dedicated PR

MalpenZibo commented:
I notice that changing the configuration (the log_level in this case) breaks in some way the compositor service (I don't see the workspaces and window title)

clotodex commented:
Okay - My hunch after debugging is that iced does not rewire the new subscription - for some reason, idk why. But to the best of my knowledge this "issue" is either within iced or a misuse of iced (which btw is way to easy...).
What I am noticing is that currently in the codebase, existing services are not being recreated in app.rs on a config change, but just receive an update message with the new config instead. @MalpenZibo I assume there was a reason to this. I would go ahead and implement the same for the affected modules to fix this bug.
